### ISSUE TITLE
Off by one errors

### ISSUE THREAD

**Issue by cdshaffer** [cdshaffer/starterator#9]
OK, I had originally thought that when starterator printed a single start in the suggested start like this: (5, 2345) that these positions were 1 based and therefore matched user expectations and DNAM and GBrowse  and that when there was no suggested start and starterator was printing the whole list like [(8, 13567), (28, 13771), (29, 13774), (30, 13777), (32, 13783), (33, 13801), (37, 13921)] that there were standard 0-based python counting and were therefore "off by one". However I have run across at least one situation with phage Abdiel_draft, where some genes are off by one and others are not. 

compared to DNAM, all orfs validated:
gene 1 negative strand gene, DNAM 457 - 633; starterator suggests 634
gene 29 positive strand gene, DNAM 22079; starterator suggests 22080
gene 33 negative strand gene DNAM 26250, starterator suggests 26251
gene 34 positive strand gene DNAM 26630, starterator suggests 26631
gene 36 negative strand gene DNAM 27799, startarator suggests 27800
gene 38 and all subsequent genes are off by one

it kind of looks like the problem started around gene 29 but a few genes between 29 and 37 are good.

**Comment by cdshaffer** [comment]
first test is to see if I get the same error if I just run one of the problematic genes from the above list as a single phamerated gene. So run starterator on just Abdiel_draft gene 1, results do show "suggested start" is off by one error. I then cleared the "Temporary Files" folder and tried again. This time there was not the off by one error. So clear the folder and try a "whole phamerated phage" report on Abiel_draft to see if I get the same list of incorrect genes as the above list. 

running whole phage I can see that the first gene to be analyzed (pham 256, gene 11) is correct, the next gene analyzed (pham 7426, gene 47) is incorrect and is off by one on the suggested start. The third gene analyzed (pham 171, gene 50) is off by one.next one good (pham 263, gene 31).  

checked a few of the others around gene 29 where things appear to transition from good to bad.

gene          pham          start
28              2502           good
29              2856           off by one
30              9515           good
33              599             off 
35              7091           good  
36              1903           off
37              9683           good
38              5353           off

so basically the same as the initial. so really not clear why some fail and some succeed but the good news is that for debugging, the first one through corrrect, the next fails.

**Comment by cdshaffer** [comment]
ok
I am thinking of adding new instance vaiables to the phamgene.py PhamGene which is of type gene.

looking at this code it looks like both Gene and PhamGene have instance variables called start and stop. I wonder if this is the source of some of the issues with some genes having errors and some genes not having errors.

**Comment by cdshaffer** [comment]
I need a better phage to test than abdiel as the first few phams are quite large so it takes a while to do any tests. So looking for phage where the first few genes are done quickly and we have some correct and some off by 1. So phage marvin is the first of the phage i have found where the first couple of genes analyzed are done quickly. in the analysis of marvin analysis appears in this order:
gene                    off by 1?
59959 rev                no
61144 for                 no
19043 For                no
46218 For              yes -but its the long list so expected

this ran much faster to get to the 1st off by one than using abdiel. Try results for phage Dori

16796 for OK
62684Rev off by one

ok this might be worth poking at.

**Comment by cdshaffer** [comment]
I have now realized there is more than one issue with off by one. First all reverse genes where the annotated start **is** the most choosen start in the whole pham are off by one. This does not explain the positive strand genes but I have added code to only add one to gene.start if the gene is reverse when calculating the suggested start.

**Comment by cdshaffer** [comment]
I have run a double check with phage abdiel and all the off by one errors in that phage  are corrected. The lists still have zero based counting but the single "suggested starts" now report coordinates in 1 based counting. I will call this one closed for now.
